### ISSUE TITLE
Softmax tempering implementation in neural-admixture

### ISSUE THREAD

**Issue by rochevin** [AI-sandbox/neural-admixture#45]
Hello,

Thank you for providing such an easy-to-use package as a great alternative to ADMIXTURE.

I'm working on a dataset where I’d like to explicitly control the degree of admixture, and I think softmax tempering would be useful. In your paper, you mention two mechanisms for controlling softness:

- The regularization parameter during training, and in particular, playing with the $\lambda$ parameter to adjust the softness.
- The logit scaling $\tau$ to control the softness level before applying softmax operation and cluster assignment.

However, I couldn't see any $\lambda$ or $\tau$ implemented in your package. Could you confirm whether these options are implemented?

From what I understand, $\tau$ could be applied manually by extracting the logits (from `Q_P.multihead_encoder`), scaling them by $\tau$, and then applying `model.softmax` per head. Is that correct?

But for $\lambda$, I don’t see any regularization term added in the optimizer or training step:

- [`self.optimizer = self.raw_model.create_custom_adam(device=self.device, lr=self.lr)`](https://github.com/AI-sandbox/neural-admixture/blob/main/neural_admixture/model/neural_admixture.py#L363C9-L363C91)
- or during [`_run_step`](https://github.com/AI-sandbox/neural-admixture/blob/main/neural_admixture/model/neural_admixture.py#L419C9-L419C18)

Perhaps I missed it, but it seems there’s no way to control $\lambda$ as an exposed parameter. Could you clarify how $\lambda$ should be set, or confirm if it's unsupported ? 

Thanks in advance for your help,
Kind regards,
Vincent Rocher

**Comment by joansaurina** [comment]
Hi @rochevin,

Thank you for your kind words and for taking the time to explore our package — we’re glad to hear you find it a useful alternative to ADMIXTURE.

You’re correct that the current version does not include explicit support for the λ regularization term or the τ logit scaling parameter. Our algorithm is currently designed to infer the underlying structure of the SNP data as faithfully as possible, without additional softmax tempering controls.

That said, we appreciate your suggestion — adding options for λ and τ could provide valuable flexibility for users who want to explore different degrees of admixture smoothness. We’ll consider including these features in a future release if there’s community interest.

Thank you again for the thoughtful feedback!

Best regards,
Joan

**Comment by rochevin** [comment]
Thanks a lot for your quick response. 

It could also be helpful to add a short section in the README explaining which features from the paper are not implemented yet. That would make it easier for future users to understand the current capabilities and avoid confusion.
